cli/internal/encoding/gitoutput/gitoutput.go

```diff
@@ -0,0 +1,270 @@
+	ObjectType field = 2
+	// ObjectName is the 40-character SHA hash
+	ObjectName field = 3
+	// ObjectStage is a value 0-3.
+	ObjectStage field = 4
+	// StatusX is the first character of the two-character output from `git status`.
+	StatusX field = 5
+	// StatusY is the second character of the two-character output from `git status`.
+	StatusY field = 6
+	// Path is the file path under version control in `git`.
+	Path field = 7
+)
+
+// Separators that appear in the output of `git`
+const space rune = ' '
+const tab rune = '\t'
+const nul rune = '\000'
+
+// A ParseError is returned for parsing errors.
+// Entries and columns are both 1-indexed.
+type ParseError struct {
+	Entry  int   // Entry where the error occurred
+	Column int   // Column where the error occurred
+	Err    error // The actual error
+}
+
+func (e *ParseError) Error() string {
+	return fmt.Sprintf("parse error on entry %d, column %d: %v", e.Entry, e.Column, e.Err)
+}
+
+func (e *ParseError) Unwrap() error { return e.Err }
+
+// These are the errors that can be returned in ParseError.Err.
+var (
+	ErrInvalidObjectMode    = errors.New("object mode is not valid")
+	ErrInvalidObjectType    = errors.New("object type is not valid")
+	ErrInvalidObjectName    = errors.New("object name is not valid")
+	ErrInvalidObjectStage   = errors.New("object stage is not valid")
+	ErrInvalidObjectStatusX = errors.New("object status x is not valid")
+	ErrInvalidObjectStatusY = errors.New("object status y is not valid")
+	ErrInvalidPath          = errors.New("path is not valid")
+	ErrFieldCount           = errors.New("too many fields")
+)
+
+// A Reader reads records from `git`'s output`.
+type Reader struct {
+	// ReuseRecord controls whether calls to Read may return a slice sharing
+	// the backing array of the previous call's returned slice for performance.
+	// By default, each call to Read returns newly allocated memory owned by the caller.
+	ReuseRecord bool
+
+	// Fields specifies the type of each field.
+	Fields []field
+
+	reader *bufio.Reader
+
+	// numEntry is the current entry being read in the `git` output.
+	numEntry int
+
+	// rawBuffer is an entry buffer only used by the readEntry method.
+	rawBuffer []byte
+
+	// recordBuffer holds the unescaped fields, one after another.
+	// The fields can be accessed by using the indexes in fieldIndexes.
+	recordBuffer []byte
+
+	// fieldIndexes is an index of fields inside recordBuffer.
+	// The i'th field ends at offset fieldIndexes[i] in recordBuffer.
+	fieldIndexes []int
+
+	// fieldPositions is an index of field positions for the
+	// last record returned by Read.
+	fieldPositions []position
+
+	// lastRecord is a record cache and only used when ReuseRecord == true.
+	lastRecord []string
+}
+
+// NewLSTreeReader returns a new Reader that reads from reader.
+func NewLSTreeReader(reader io.Reader) *Reader {
+	return &Reader{
+		reader: bufio.NewReader(reader),
+		Fields: []field{ObjectMode, ObjectType, ObjectName, Path},
```

Comment:
These `Fields` slices should probably be constants (eh, `var` I guess). There might even be a slick way to package them along with some functions to read the appropriate field to get hash and path.

Reply:
At the end of everything if I do something to turn the `[][]string` to a type I can do this; great idea.